frame/contracts/src/exec.rs

```diff
@@ -1021,7 +1083,12 @@ where
 	}
 
 	fn caller(&self) -> &T::AccountId {
-		self.frames().nth(1).map(|f| &f.account_id).unwrap_or(&self.origin)
+		let frame = if self.frames().collect::<Vec<_>>().len() <= 1 {
```

Comment:
You could just check whether `delegated_call` is `Some`?

Reply:
`delegated_call` is part of the frame argument, not frame itself.
Should I add a `caller` field to the frame?

Reply:
Ahh I forgot that. Yeah it should be added to the frame. So we can easily check whether this is a delegated call and therefore has a different caller.